`kw patch-hub`: Add reliable fetch of latest patchsets from mailing list

davidbtadokoro opened the pull request:
This PR implement the most critical point in `kw patch-hub` development, which is the interaction with the Lore API to fetch the latest patchsets from a given mailing list.

It introduces a realiable way to navigate through the whole history of a mailing list.

It can also be used as a base for making more specific queries to lore archives.

Commit:
src: lib: lore: Add fetch of last n patchsets from mailing list

This commit implements the fetching of the last n patchsets sent to a
given public mailing list in the lore.kernel.org archives.

The global variable `NUMBER_OF_PATCHSETS_PER_FETCH` defines the number
of patchsets needed to be fetch before displayng to the user and the
global variable `TIMEFRAME_IN_DAYS` defines the size of the timeframe
considered when querying lore.kernel.org, in other words, the size of
the period or chunk of time for each request to the server.

This is a big commit that creates a lot of functions, but, essentially,
it takes the same steps that the old `processing_new_patches` function
did. This function was renamed to `fetch_latest_patchsets_from` and the
created functions are delegations within it. These delegations make
`fetch_latest_patchsets_from` somewhat of a mediator and allow us to
test each step of what used to be done entirely by
`processing_new_patches`.

The major implementation change is the addition of a while loop that
fetches patches in adjacent chunks of time until the required number of
patchsets is obtained.

It is worth noting the lore API for fetching patches. For each public
mailing list archived, there is an URL that is common to every query for
patches. For example, any query for patches of 'amd-gfx' has to have

  https://lore.kernel.org/amd-gfx/

in its URL. Query filters can be applied to match only patches that
fulfill specific conditions.

For the purposes of this commit, we need only to filter based on time.
An example of query URL to request patches of 'amd-gfx' from 4 days ago
until 2023-01-01 00:00:00 of the 'amd-gfx' public mailing list.

  https://lore.kernel.org/amd-gfx/
  ?q=rt:4.day.ago..2023-01-01T00:00:00Z+AND+NOT+s:Re&x=A

The 'rt:' refers (seemingly) to the recieved time of the message in the
lore servers. We use this filter instead of 'd:' (date time accused by
the message in the 'Date' header), because lore responds queries with an
XML ordered by the attribute filtered by 'rt:' instead of 'd:'. As we
want to have a consistent ordered list, and, preferebly avoid the
overhead of having to order the list itself, we use 'rt:'. Note that the
whole filter is 'rt:4.day.ago..2023-01-01T00:00:00Z'.

The filter 's:Re:' matches only messages without 'Re:' in the subject,
to filter out replies, as we want to have only patches.

Closes: #799

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -312,67 +320,75 @@ function processing_new_patches()
-          "$processed_line" "$message_id_link" "$title" &
-        ((index++))
+  while [[ "$index" -lt "$NUMBER_OF_PATCHSETS_FETCHED" ]]; do
+    query_filter="?q=rt:${days}.day.ago..${last_patch_timestamp}+AND+NOT+s:Re&x=A"
```
Another possibility for `query_filter` is `"?q=s:PATCH+AND+NOT+s:Re&o=${inferior_limit}&x=A"` where the **patch** of number `inferior_limit` + 1 sent to the mailing list is the first one in the response list.

Go to https://lore.kernel.org/amd-gfx/?q=s:PATCH+AND+NOT+s:Re&o=0 and https://lore.kernel.org/amd-gfx/?q=s:PATCH+AND+NOT+s:Re&o=200 for examples of the 1-200 last **patches** sent to the amd-gfx list and of the 201-400 last **patches**, respectively.

rodrigosiqueira replied:
We need to document all of your work somewhere, something that highlights explanations of the above queries and their difference. Maybe a new kw blog post rich in details about this work.

davidbtadokoro replied:
Nice idea. This is also an opportunity to sturdy further the lore API.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -34,6 +34,14 @@ declare -gA available_lore_mailing_lists
 
+# TODO: Remove hardcode and add setting for this
+# Number of patchsets needed to fetch when querying a lore mailing list
+declare -gr NUMBER_OF_PATCHSETS_FETCHED=100
```
Let's use 30 by default.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -312,67 +320,75 @@ function processing_new_patches()
-        ((index++))
+  while [[ "$index" -lt "$NUMBER_OF_PATCHSETS_FETCHED" ]]; do
+    query_filter="?q=rt:${days}.day.ago..${last_patch_timestamp}+AND+NOT+s:Re&x=A"
+    query_url="${LORE_URL}/${target_mailing_list}/${query_filter}"
```
I was wondering if we should create a new function to build the URL and ensure it is valid inside this new function (e.g., `compose_url_with_verification()`).

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -312,67 +320,75 @@ function processing_new_patches()
+    query_filter="?q=rt:${days}.day.ago..${last_patch_timestamp}+AND+NOT+s:Re&x=A"
+    query_url="${LORE_URL}/${target_mailing_list}/${query_filter}"
+
+    download "$query_url" "$list_patches_file_name" "$CACHE_LORE_DIR" "$flag" || return "$?"
```
Since this is a large line, I'm under the impression that using a normal `if` will make this code easier to read.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -312,67 +320,75 @@ function processing_new_patches()
+
+    download "$query_url" "$list_patches_file_name" "$CACHE_LORE_DIR" "$flag" || return "$?"
+
+    xpath_query="${NAME_EXP}|${EMAIL_EXP}|${TITLE_EXP}|${LINK_EXP}"
+    raw_xml=$(< "${CACHE_LORE_DIR}/${list_patches_file_name}")
+    pre_processed=$(printf '%s' "$raw_xml" | xpath -q -e "$xpath_query")
```
This is another code that might be better if it has a function.

P.s.: Document the use of XPath in the lore context; it will also be good.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -312,67 +320,75 @@ function processing_new_patches()
+
+    # Converting to:
+    #  Author, email, version, total patches, patch title, link
+    shared_dir_for_parallelism=$(mktemp -d)
```
If you want to avoid access to the file system, you can try to replace `mktemp -d` with `/dev/shm` since this is basically a folder in the RAM system. Before you try this, I recommend you check how fast you will use/dev/shm compared with /tmp to ensure that accessing the disk is part of the bottleneck.

davidbtadokoro replied:
Thanks a lot for the idea. This may help with the performance issues. Anyway, going to keep this in mind for when we optimize the feature. Thanks again!

davidbtadokoro commented:
Hi @rodrigosiqueira.
Thanks a lot for the many invaluable feedbacks and comments. Going to refine this draft ASAP.

davidbtadokoro commented:
Hi, @rodrigosiqueira.
I updated the commit with the changes you commented and some more. Sorry for the big commit.
Also, I'm working on a commit to follow it that adds pagination (i.e., the user can see the N older/earlier patchsets).

davidbtadokoro commented:
Really strange, the CI/CD fails because the `xpath` command is not found.

![Screen Shot 2023-08-10 at 00 34 11](https://github.com/kworkflow/kworkflow/assets/95055784/9bf9cfbe-e079-4efe-89ba-694f054dab38)

Tried some fixes but with no avail 😞

Any ideas?

rodrigosiqueira commented:
> Really strange, the CI/CD fails because the `xpath` command is not found.
> 
> ![Screen Shot 2023-08-10 at 00 34 11](https://user-images.githubusercontent.com/95055784/259614259-9bf9cfbe-e079-4efe-89ba-694f054dab38.png)
> 
> The test runs fine in local and it doesn't seems to be a network environment issue in the runner side.
> 
> Tried some fixes but with no avail disappointed
> 
> Any ideas?

@davidbtadokoro 

If CI is missing some package, check https://github.com/kworkflow/kworkflow/blob/unstable/.github/workflows/unit_tests.yml . I guess you probably need to update it.

davidbtadokoro commented:
> > Really strange, the CI/CD fails because the `xpath` command is not found.
> > 
> > ![Screen Shot 2023-08-10 at 00 34 11](https://user-images.githubusercontent.com/95055784/259614259-9bf9cfbe-e079-4efe-89ba-694f054dab38.png)
> > 
> > The test runs fine in local and it doesn't seems to be a network environment issue in the runner side.
> > 
> > Tried some fixes but with no avail disappointed
> > 
> > Any ideas?
> 
> @davidbtadokoro 
> 
> If CI is missing some package, check https://github.com/kworkflow/kworkflow/blob/unstable/.github/workflows/unit_tests.yml . I guess you probably need to update it.

Oh, that must be it! I tried changing https://github.com/kworkflow/kworkflow/blob/master/.github/workflows/test_setup_and_docs.yml instead of unit_tests.yml ;)

davidbtadokoro commented:
> I ran a quick test on `ath10k`, and I noticed it took a long time. Is it possible to let the user know that some lists will take more time? Also, I really want to see `next` and `previous` pages as a way to validate your new approach.

For sure, I can add a message for it. But I think that the performance issue (besides the fact that we may need to add some optimizations in the future and that the list may be less "busy" than others) may be influenced by some patches that are discarded by `kw patch-hub` ([draft issue](https://github.com/orgs/kworkflow/projects/2/views/1?pane=issue&itemId=35278244)).

And sorry for the delay on the `next` and `previous` pages. I'm working on it and it is looking good.

Commit:
src: ui: patch_hub: patch_hub_core: Extract `show_new_patches_in_the_mailing_list`

This commit extracts the function `show_new_patches_in_the_mailing_list`
defined in `src/ui/patch_hub/patch_hub_core.sh` to the new file
`src/ui/patch_hub/latest_patchsets.sh`. The unit test related to this
function was also extracted to a new file. This change only moves code
around and doesn't change any implementation to avoid introducing
sneaky regressions.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: latest_patchsets_from_mailing_list: Refactor screen and function names

To conform with the pattern of screen and function names in the `kw
patch-hub` feature, this commit changes the name of the screen
responsible for showing the latest patchsets from a given lore mailing
list from `show_new_patches_in_the_mailing_list` to
`latest_patchsets_from_mailing_list`, and its handler function from
`show_new_patches_in_the_mailing_list` to
`show_latest_patchsets_from_mailing_list`.

This commit also adds a superficial documentation to the
`show_latest_patchsets_from_mailing_list` function.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: latest_patchsets_from_mailing_list: Add pagination with incremental fetches

The `kw patch-hub` feature can display the latest N patchsets from
a given lore public mailing list to the user. However, the if the user
wants to see the (N+1)th patchset, he/she has to change the source code
(to alter the hardcoded variable that determines the number of
patchsets), launch the feature again, and fetch and process the
N patchsets that were just displayed plus the (N+1)th patchset. Also,
currently the N patchsets are displayed in a single dialog screen.

This is really impractical, so this commit adds pagination for the
fetch and processment of patchsets. Essentially, now the user can see
the N patchsets broken in many dialog screens (pages) with M patchsets
each, and, more importantly, if the user wants to see an older page from
the oldest fetched and processed page, only this (and maybe more) page
will be fetched and processed.

This change allows the user to potentially go down all the history of
patchsets from a given lore mailing list. Technically the limit of
patchsets that can be displayed is the maximum size of a Bash array, but
this data structure seems to not have a limit.

Note: Although navigating through loaded pages is fast enough, page
loadings are really slow (from many seconds to minutes) and inconsistent
accross different mailing lists. This is a combination of needed
optimizations and corrections to the code that should be addressed to
allow for a better experience to the user.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -276,72 +287,151 @@ function reset_list_of_mailinglist_patches()
+}
+
+# This function pre-processes an XML file containing a list of patches, extracting
+# just the metadata needed to process a XML element representing a patch. The `xpath`
```
/a XML/an XML/

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -276,72 +287,151 @@ function reset_list_of_mailinglist_patches()
+# This function pre-processes an XML file containing a list of patches, extracting
+# just the metadata needed to process a XML element representing a patch. The `xpath`
+# command is used to capture the desired fields for each patch. A simplified example
+# of a XML element representing a patch is:
```
/a XML/an XML/

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -276,72 +287,151 @@ function reset_list_of_mailinglist_patches()
+# command is used to capture the desired fields for each patch. A simplified example
+# of a XML element representing a patch is:
+#   <entry>
+#	  	<author>
+#	  		<name>Jay Cornwall</name>
+#	  		<email>jay.cornwall@amd.com</email>
+#	  	</author>
+#	  	<title>[PATCH] drm/amdkfd: Add missing tba_hi programming on aldebaran</title>
+#	  	<updated>2023-08-09T21:27:00Z</updated>
+#	  	<link href="http://lore.kernel.org/amd-gfx/20230809212615.137674-1-jay.cornwall@amd.com/"/>
+#	  	<id></id>
+#	  	<content></content>
+#	  </entry>
```
Reduce the indentation to align with `<entry> `

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -352,29 +442,102 @@ function processing_new_patches()
+
+    # Processing patches into patchsets that will be stored in `list_of_mailinglist_patches`.
+    pre_processed_patches=$(pre_process_xml_result "$raw_xml")
+    # TODO: Is passing `pre_processed_patches` (huge string) as argument a possible bottleneck?
```
What do you consider huge?

davidbtadokoro replied:
Don't know if this is huge, but in the ten thousands lines and one hundred thousands words mark.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -352,29 +442,102 @@ function processing_new_patches()
+    # Update lower and upper ends of time period to query.
+    days="$((days + TIMEFRAME_IN_DAYS))"
+    last_patch_timestamp=$(printf '%s' "$raw_xml" | xpath -q -e '//entry[last()]/updated/text()')
+    last_patch_timestamp=$(TZ=UTC date --date "${last_patch_timestamp} - 1 seconds" '+%Y-%m-%dT%H:%M:%SZ')
```
Do we have any helpers for that date manipulation?

rodrigosiqueira commented on src/lib/web.sh:
```diff
@@ -49,3 +49,26 @@ function replace_http_by_https()
+
+  [[ ! -f "$file_path" ]] && return 2 # ENOENT
+
+  first_line_of_file=$(head -n 1 "$file_path" | tr '[:upper:]' '[:lower:]')
+  if [[ "$first_line_of_file" =~ ^(<html|<\!doctype html>) ]]; then
```
This is usually true; I guess this will always work for the lore use case 100%. However, in terms of URLs, this is not always true. I would add more checks if the first check fails to make this function more robust. For example, you can look for `<header>` and `<body>`; if this function does not find anything like that, then you return false.

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -440,4 +441,114 @@ function test_save_new_lore_config()
   assert_equals_helper 'Wrong lore.config contents' "$LINENO" "$expected" "$output"
 }
 
+function test_compose_lore_query_url_with_verification()
```
I suggest you split these tests by cases. For example, move the invalid option to its own test and the good cases as well.

rodrigosiqueira commented on tests/lib/web_test.sh:
```diff
@@ -50,4 +73,35 @@ function test_replace_http_by_https()
   assert_equals_helper 'Expected https' "($LINENO)" "$output" "$expected"
 }
 
+function test_is_html_file()
```
split this test. See my previous comment about that.

rodrigosiqueira commented on .github/workflows/kcov.yml:
```diff
@@ -20,7 +20,7 @@ jobs:
       - name: Install dependencies
         run: |
-          sudo apt install -y kcov shunit2 bc sqlite3 bsdmainutils
+          sudo apt install -y kcov shunit2 bc sqlite3 bsdmainutils libxml-xpath-perl
```
You can move this change to the first commit.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -108,6 +108,85 @@ function create_menu_options()
+# variable. If the user hit the 'OK' button 0 is returned, if hit the 'Exit' button 2
+# is returned, if hit the `Previous` button 3 is returned and if hit the `Next` button 1
+# is returned.
+function create_menu_options_with_pagination()
```
Do we really need this function? This looks too specific to be part of a lib. Maybe I'm missing something but it looks like you can change `create_menu_options()` to make it more flexible to support the behavior that you implemented here.

rodrigosiqueira replied:
Additionally, after passing through this commit, I believe this change does not belong to this commit. You need to create a specific commit only for this.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -35,16 +35,24 @@ declare -gA available_lore_mailing_lists
 # TODO: Remove hardcode and add setting for this
-# Number of patchsets needed to fetch when querying a lore mailing list
-declare -gr NUMBER_OF_PATCHSETS_PER_FETCH=30
+# Number of patchsets per page when querying a lore mailing list
+declare -g NUMBER_OF_PATCHSETS_PER_PAGE=30
```
As part of this PR, I would like to see these configuration variables in the config file.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -276,72 +287,151 @@ function reset_list_of_mailinglist_patches()
+
+  # Verifying if upper end is valid, i.e., is a real date in the correct format
+  if [[ -n "$upper_end_in_timestamp" ]]; then
+    date '+%Y-%m-%dTH:M:SZ' --date "$upper_end_in_timestamp" > /dev/null 2>&1
```
Agree, we need to test this. I was thinking if this helper isn't too specific to lore, so I'm going to add it to `lore.sh`. What do you think?

davidbtadokoro replied:
Actually, do you know if the `YYYY-mm-ddTHH:MM:SSZ` is used anywhere or is a known format at all? In that case, we can add to helper to `kw_time_and_date` with a name like `is_in_<name_of_format>_timestamp_format`

Commit:
src: lib: dialog_ui: Make `create_menu_options` more flexible

The `create_menu_options` function is responsible to create a dialog
screen with a list of options from which the user can choose one.
However, there is no way to:

- Add a 'Help' button.
- Customize the 'Extra' button label.
- Choose only a range of options from the array passed as argument to be
  displayed.

This commit extends the function to address these points above, making
it more flexible to, for example, create a menu with pagination.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
Hey @rodrigosiqueira, updated the PR, but I'm still working on the last commit to add the settings to the config file (and maybe to the 'Settings' screen). Doing it ASAP.

Commit:
src: lib: dialog_ui: Add function to create rangebox screen

This commit adds a function to create rangebox screens using dialog.
A rangebox is a bar, much like a volume bar, with a minimum, maximum,
and starting value that the user can decrement or increment with the UP
and DOWN arrow keys. This allows us to guaranteedly capture an integer
from the user from a restricted range.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch-hub: settings: Add 'Patchsets Per Page' and 'Lore Requests Timeframe' screens

For the fetching of the latest patchsets from a target mailing list,
patch-hub uses two global variables declared in `src/lib/lore.sh`
`TIMEFRAME_SIZE_IN_DAYS` and `NUMBER_OF_PATCHSETS_PER_PAGE`. These
relate to the time period considered when fetching data from lore
servers and the number of patchsets to be displayed when listing
patchsets and their combination influences in loading times as well as
cutomization. Also, for each mailing list (because of how busy or idle
it is) there may be a optimal combination.

For this purpose, this commit extracts these two global variables from
`src/lib/lore.sh` to `lore.config` as settings. These can be changed via
directly editing the file or via the patch-hub feature through the
'Dashboard'->'Settings'->'Patchsets Per Page' and
'Dashboard'->'Settings'->'Lore Requests Timeframe' screens added.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: dialog_ui: Support indexed array in `create_choice_list_screen`

The `create_choice_list_screen` displays a dialog screen with a list of
choices for the user to select one. The choices are passed as argument
as an associative array with pairs 'choice' and 'choice description'.
The problem with this is, as Bash associative arrays are hashed, there
is no way to preserve the order of definition when looping through its
elements.

As sometimes we want the order of definition of the choices to be
consistent and the description is not really necessary, this commit adds
support for indexed arrays in `create_choice_list_screen`. Indexed
arrays preserve the order of definition when looping through its
elements (in a foreach, for example), which allow us to create a choice
list with a defined display order.

This commit doesn't alters the signature of `create_choice_list_screen`,
it only allows the argument `_choices` (a Bash namereference) to be an
indexed array, instead of restricting it to an associative array, by
detecting which type was passed and taking the according action.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: lib: lore: Change patchsets fetch strategy

By testing the lore API, I noticed a behaviour: lore responses to
queries are paginated and the maximum number of messages per response is
200. For example, if we make a query to a public mailing list archived
on lore and there are, say, 750 messages matched, the first response
would contain only messages 1 to 200. If we want to get the next chunk
(201 to 400) we need to add the parameter `o=200` in the query string,
which is the same as clicking on the button 'next (older)' when
accessing lore.kernel.org on the browser.

`kw patch-hub` patchset fetch strategy is to consider adjacent chunks of
time (timeframes), by making query strings to match patchsets aggainst
these chunks. However, if the number of messages matched is greater than
200, messages 201 onwards won't be present in the response, which means
that the same query with a smaller chunk of time renders the same
results.

Although, by the way the implementation was done, we don't have
a correctness problem, all the date manipulation that is made to manage
these chunks of time and keeping them consistent is unnecessary, adds
more complexity to the code, and maybe results in a performance
overhead.

With all this in mind, this commit changes the patchsets fetch strategy
by completely removing any date manipulation and moving the focus to
manipulate only the `o` parameter of lore API. The core procedures of
the strategy are mantained, because the change entails only this
"translation" from chunks of time to chunks of messages.

Most changes are contained in `src/lib/lore.sh` (the Model), but some
collateral updates are done around `kw patch-hub`, in special the
removal of the 'Lore Requests Timeframe' setting.

Note: This new strategy has a possible "unpleasant behaviour" that is,
if the list gets updated during a fetch session, patchsets may appear
duplicated, but without messing the order or any functionality. In other
words, when a new message to a given list is recieved on lore, message
"number 1" becomes "number 2" and so on, so when requesting the next
chunk of messages from lore, message 200 (that was already processed)
may be processed again as message 201. This can be considered a bug that
doesn't break the feature in any way, but can be a hassle if the list
gets updated too frequently (imagine having to race aggainst the
updates).

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -55,52 +65,60 @@ function create_menu_options()
   local cmd
 
-  if [[ "${#_menu_list_string_array[@]}" -eq 0 ]]; then
+  if [[ "${#_options_array[@]}" -eq 0 ]]; then
```
How about changing `_options_array` to `_list_of_options_array`?

rodrigosiqueira commented on src/ui/patch_hub/latest_patchsets_from_mailing_list.sh:
```diff
@@ -1,24 +1,58 @@
+  local box_title
+  local extra_label='Previous'
+
+  create_loading_screen_notification "Loading patchsets from ${current_mailing_list} list (page ${PAGE})"
```
I'm unsure if I like the `page ${PAGE}` part... to me, it is a little bit weird that we see "page 1" and after two next, we see "page 3". If we can show something like "pages 1-2", it should be fine. If it is not possible, maybe we should not display any number to avoid misleading users.

davidbtadokoro replied:
Hmm, you are right. But I don't know if we can know in advance the number of patchsets fetched, as this number fluctuates across queries. Think the best approach for now is to not display the number.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -433,17 +437,30 @@ function create_choice_list_screen()
+  if [[ "$(typeset -p "$3")" =~ 'declare -a' ]]; then
+    for choice in "${_choices[@]}"; do
+      choice=$(str_escape_single_quotes "$choice")
+      if [[ "${_check_statuses["$index"]}" == 1 ]]; then
```
I believe that the reason that the default option was not working correctly came from the fact that you mapped one array (`_check_statuses`) against an associative array (`_choices`), and associative arrays do not follow the same order as a regular array. Before your change, you were trying to use an array (index) to match an associative array order that can not be guaranteed. If you change `_check_status` to an associative array, and require the user to use the same key for `_check_statuses` and `_choices`, you can map the default options.

davidbtadokoro replied:
Sorry for the messy commit message, but the issue wasn't that `_choices` and `_check_statuses` were not in sync, as I make the caller populate `_check_statuses` in the same manner that we do the matching in this function, i.e., with a foreach loop which has a consistent order when looping through keys of an associative array across multiple "calls". The problem was that this order of keys isn't the same as the order of declaration, neither the lexicographic order, like you can see in this screenshot:

<img width="1059" alt="Screen Shot 2023-08-20 at 18 34 34" src="https://github.com/kworkflow/kworkflow/assets/95055784/62060649-acaa-4404-92e3-0e6e564c2461">

This made it virtually impossible for the caller to define an order of displaying the choices with an associative array, which was the problem that I tried tackling with this commit.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -318,13 +310,16 @@ function process_name()
   list_of_mailinglist_patches=()
   PATCHSETS_PROCESSED=0
-  DAYS="$TIMEFRAME_SIZE_IN_DAYS"
+  DAYS="$lore_requests_timeframe"
```
btw, you can still have a fallback option in case we have any parser issue or if the user set it to a weird value. I.e., set a default case if this variable is invalid.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -276,72 +312,146 @@ function reset_list_of_mailinglist_patches()
+#   Jay Cornwall
+#   jay.cornwall@amd.com
+#   [PATCH] drm/amdkfd: Add missing tba_hi programming on aldebaran
+#    href="http://lore.kernel.org/amd-gfx/20230809212615.137674-1-jay.cornwall@amd.com/"
```
Yes, `//entry/link/@href` in the xpath query results in this leading space. We can remove it, but don't know if this extra processing is worth (maybe I'm underthinking it, but it seems that we would have to add another `sed` in the pipe). What do you think?

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -276,72 +312,146 @@ function reset_list_of_mailinglist_patches()
+{
+  local pre_processed_patches="$1"
+  local shared_dir_for_parallelism
+  local starting_index
+  local line
+  local processed_patchset
+  local patch_url
+  local patch_title
+  local count
```
Cool! Didn't know this pattern of declaration.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -276,72 +312,146 @@ function reset_list_of_mailinglist_patches()
-        ((index++))
+        thread_for_process_patch "$PATCHSETS_PROCESSED" "$shared_dir_for_parallelism" \
+          "$processed_patchset" "$patch_url" "$patch_title" &
+        ((PATCHSETS_PROCESSED++))
```
In a later commitl, this becomes more evident, but my intetion was to use this global variable to hold the "state" across many calls of `fetch_latest_patchsets_from` (and its subroutines). We could shift the responsability to the caller (`PATCHSETS_PROCESSED` becomes an argument for `fetch_latest_patchsets_from` and its subroutines), but I thought this broke the MVC pattern that is loosely implemented and because I think the caller would have to use a global variable also.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -352,29 +462,102 @@ function processing_new_patches()
+    # Building URL for querying lore servers for a xml file with patches.
+    lore_query_url=$(compose_lore_query_url_with_verification "$target_mailing_list" "$days" "$last_patch_timestamp")
+    ret="$?"
+    [[ "$ret" != 0 ]] && return "$ret"
```
`compose_lore_query_url_with_verification` returning something other than 0 means that either one of its three arguments is broken (empty or invalid), which would happen in any next iteration of the while loop. Maybe `fetch_latest_patchsets_from` "knowing" this is a sign of high coupling, though...

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -352,29 +462,102 @@ function processing_new_patches()
+
+    # Request xml file with patches.
+    download "$lore_query_url" "$xml_result_file_name" "$CACHE_LORE_DIR" "$flag"
+    ret="$?"
+    [[ "$ret" != 0 ]] && return "$ret"
```
In the same line of my last response, a bad URL means bad URLs for any of the next iterations of the while loop. However, I agree that if a download failed because of network issues, for example, we should retry to download it. Do you have an idea to fine grain those errors to treat them accordingly?

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -352,29 +462,102 @@ function processing_new_patches()
+    raw_xml=$(< "${CACHE_LORE_DIR}/${xml_result_file_name}")
+
+    # Processing patches into patchsets that will be stored in `list_of_mailinglist_patches`.
+    pre_processed_patches=$(pre_process_xml_result "$raw_xml")
```
I was wrong in my previous guess of the size of `raw_xml`. At most, it has 800 lines, relating to the maximum 200 patches in lore's response, but I do agree on passing the path instead.

rodrigosiqueira commented:
Thanks for this improvement. Change merged into the unstable branch.
